design/loadbalancer-hash-policy-design.md

```diff
@@ -0,0 +1,185 @@
+The [`WeightedLeastRequest` option](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#weighted-least-request) uses different algorithms to route requests to upstream hosts based on the relative weights and the number of active requests to that instance.
+Envoy also provides a [ring hash](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/load_balancers#ring-hash) load balancing policy which hashes some property of a request in order to select an upstream host.
+Envoy produces a consistent hash based on an [attribute of a request](https://www.envoyproxy.io/docs/envoy/v1.17.0/api-v3/config/route/v3/route_components.proto#config-route-v3-routeaction-hashpolicy) to allow clients to effectively select a consistent backend instance requests will be sent to.
+Contour currently only exposes the [cookie value](https://www.envoyproxy.io/docs/envoy/v1.17.0/api-v3/config/route/v3/route_components.proto#config-route-v3-routeaction-hashpolicy-cookie) based flavor as the `Cookie` load balancing strategy in order to give an out of the box experience that provides the ability to implement [session affinity](session-affinity.md) to a particular upstream host.
+
+In order to cater to more advanced load balancing use cases or offer more granular control, Contour can provide additional mechanisms for configuring Envoy's load balancing hash policy.
+Specifically, [hashing HTTP request headers](https://github.com/projectcontour/contour/issues/3099) can help to route client requests to a service instance, a feature that may be useful for stateful applications being transformed to run in Kubernetes or monolithic applications in the process of being split into microservices.
+A client may send a consistent identifier in a header value in order to specify the same upstream service instance should handle all requests with that specific value.
+In addition, [supporting multiple hash policies](https://github.com/projectcontour/contour/issues/3044) can give even more flexibility.
+If a particular load balancing attribute is not available on a request, users can specify fallback policies on additional attributes to still segment traffic between backends, rather than having Envoy continue on to route requests using the round-robin or random mechanisms.
+
+## Goals
+- Contour should offer the ability to load balance requests based on HTTP user specified request headers
+- The design should offer the ability to configure multiple hash policies (of which specific new types may be implemented separately from this design)
+
+## Non Goals
+- Provide advanced load balancing configuration for `ExtensionService` or `TCPRoute`
+- Provide users direct access to Envoy configuration fields/structures
+
+## High-Level Design
```

Comment:
Based on background, I assume the plan is to support the `RingHash` load balancer? Do we want to support `Maglev` as well, and if so, how would we do that?

Reply:
`Maglev` is a drop-in replacement for `RingHash` as far as the Envoy config is concerned. So I think we can kick that can down the road.

Reply:
Yeah, could be another "strategy" to use MagLev, instead of Ring Hash, could lead us to change the name of the strategy proposed here or just add a new one in the future if people ask for MagLev specifically

Reply:
Yeah I don't think we need to add it now. Should be fine to add `Maglev` as a strategy in the future.